core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -570,6 +423,190 @@ class TiBatchWrite(@transient val df: DataFrame,
+      allocatedRdd = rdd.zipWithIndex.map {
+        case (data, i) =>
+          val handle = i + start
+          (data, handle)
+      }
+    } else {
+      allocatedRdd = rdd.map { data =>
+        val handle =
+          data.row.get(handleCol.getOffset, handleCol.getType).asInstanceOf[Number].longValue()
+        (data, handle)
+      }
+    }
+
+    val rowKeyAllocatedRdd = allocatedRdd.map {
+      case (data, handle) =>
+        (
+          new SerializableKey(RowKey.toRowKey(tiTableInfo.getId, handle).getBytes),
+          encodeTiRow(data.row)
+        )
+    }
+
+    val indexKeyAllocatedRdd = allocatedRdd
+      .map {
+        case (data, handle) =>
+          // extract index key from keyWithDupInfo
+          val indexKeys = data.indexKeys.map { key =>
+            key.key
+          }
+          indexKeys.map { key =>
+            val cdo = new CodecDataOutput()
+            cdo.writeLong(handle)
+            (key, cdo.toBytes)
+          }
+      }
+      .flatMap(identity)
+
+    rowKeyAllocatedRdd.union(indexKeyAllocatedRdd)
+  }
+
+  private def checkConflictWithInsertKeys(
+    rdd: RDD[toBeCheckedRow]
+  ) = {
+    if (handleCol != null) {
+      // data is conflict with TiKV
+      val handleConflict = !rdd
+        .filter {
+          // check handle column and unique indices
+          data =>
+            data.handleKey.oldRow != null
+        }
+        .isEmpty()
+
+      if (handleConflict) {
+        throw new TiBatchWriteException("data to be inserted is conflict on primary key")
+      }
+
+      val handleDuplicated = rdd
+        .map { row =>
+          (row.handleKey, 0)
+        }
+        .groupByKey()
+        .flatMap {
+          case (_, iterable) =>
+            if (iterable.size > 1) {
+              Some(iterable.size)
+            } else {
+              None
+            }
+        }
+
+      if (!handleDuplicated.isEmpty) {
+        throw new TiBatchWriteException("data to be inserted is conflict on primary key")
+      }
+    }
+
+    if (uniqueIndices.nonEmpty) {
+      val handleConflict = !rdd
+        .filter {
+          // check handle column and unique indices
+          data =>
+            data.indexKeys.exists(p => p.oldHandle != -1)
+        }
+        .isEmpty()
+
+      if (handleConflict) {
+        throw new TiBatchWriteException("data to be inserted is conflict on index key")
+      }
+      val indexConflict = rdd
+        .flatMap { row =>
+          row.indexKeys.map { key =>
+            (key, row)
+          }
+        }
+        .groupByKey()
+        .flatMap {
+          case (_, iterable) =>
+            if (iterable.size > 1) {
+              Some(iterable.size)
+            } else {
+              None
+            }
+        }
+
+      if (!indexConflict.isEmpty) {
+        throw new TiBatchWriteException("data to be inserted is conflict on unique index")
+      }
+    }
+
+    rdd
+  }
+
+  private def getKeysNeedCheck(
+    rdd: RDD[TiRow]
+  ) = {
```

Comment:
better to add `return type` for others to read

Reply:
ok